zebra/zebra_nb_state.c

```diff
@@ -155,235 +170,414 @@ lib_interface_zebra_state_mcast_group_get_elem(const char *xpath,
+			zrt = RB_NEXT(zebra_router_table_head, zrt);
+	}
+
+	if (IS_ZEBRA_DEBUG_EVENT && zrt)
+		zlog_debug("%s: vrf %s afi %s safi %s", __func__, vrf->name,
+			   afi2str(zrt->afi), safi2str(zrt->safi));
+
+	return zrt;
 }
 
-int lib_vrf_ribs_rib_get_keys(const void *list_entry,
-			      struct yang_list_keys *keys)
+int lib_vrf_zebra_ribs_rib_get_keys(const void *list_entry,
+				    struct yang_list_keys *keys)
 {
-	/* TODO: implement me. */
+	const struct zebra_router_table *zrt = list_entry;
+
+	assert(zrt);
+
+	keys->num = 2;
+	zebra_extract_afi_safi_key(keys->key[0], sizeof(keys->key[0]), zrt->afi,
+				   zrt->safi, "frr-zebra");
+
+	snprintf(keys->key[1], sizeof(keys->key[1]), "%" PRIu32, zrt->tableid);
+
+	if (IS_ZEBRA_DEBUG_EVENT)
+		zlog_debug("%s: key[0] %s key[1] %s", __func__, keys->key[0],
+			   keys->key[1]);
+
 	return NB_OK;
 }
 
-const void *lib_vrf_ribs_rib_lookup_entry(const void *parent_list_entry,
-					  const struct yang_list_keys *keys)
+const void *
+lib_vrf_zebra_ribs_rib_lookup_entry(const void *parent_list_entry,
+				    const struct yang_list_keys *keys)
 {
-	/* TODO: implement me. */
-	return NULL;
+	struct vrf *vrf = (struct vrf *)parent_list_entry;
+	struct zebra_vrf *zvrf;
+	afi_t afi = AFI_IP;
+	safi_t safi = SAFI_UNICAST;
+
+	zvrf = zebra_vrf_lookup_by_id(vrf->vrf_id);
+
+	return zebra_router_find_zrt(zvrf, zvrf->table_id, afi, safi);
 }
 
 /*
- * XPath: /frr-vrf:lib/vrf/frr-zebra:ribs/rib/route
+ * XPath: /frr-vrf:lib/vrf/frr-zebra:zebra/ribs/rib/route
  */
-const void *lib_vrf_ribs_rib_route_get_next(const void *parent_list_entry,
-					    const void *list_entry)
+const void *lib_vrf_zebra_ribs_rib_route_get_next(const void *parent_list_entry,
+						  const void *list_entry)
 {
-	/* TODO: implement me. */
-	return NULL;
+	const struct zebra_router_table *zrt = parent_list_entry;
+	const struct route_node *rn = list_entry;
+
+	if (list_entry == NULL)
+		rn = route_top(zrt->table);
+	else
+		rn = srcdest_route_next((struct route_node *)rn);
+
+	return rn;
 }
 
-int lib_vrf_ribs_rib_route_get_keys(const void *list_entry,
-				    struct yang_list_keys *keys)
+int lib_vrf_zebra_ribs_rib_route_get_keys(const void *list_entry,
+					  struct yang_list_keys *keys)
 {
-	/* TODO: implement me. */
+	const struct route_node *rn = list_entry;
+	char dst_buf[PREFIX_STRLEN] = {'\0'};
+
+	keys->num = 1;
+	prefix2str(&rn->p, dst_buf, sizeof(dst_buf));
+
+	strlcpy(keys->key[0], dst_buf, sizeof(keys->key[0]));
+
 	return NB_OK;
 }
 
 const void *
-lib_vrf_ribs_rib_route_lookup_entry(const void *parent_list_entry,
-				    const struct yang_list_keys *keys)
+lib_vrf_zebra_ribs_rib_route_lookup_entry(const void *parent_list_entry,
+					  const struct yang_list_keys *keys)
 {
-	/* TODO: implement me. */
-	return NULL;
+	const struct zebra_router_table *zrt = parent_list_entry;
+	struct prefix p;
+	struct route_node *rn;
+
+	yang_str2prefix(keys->key[0], &p);
+
+	rn = route_node_match(zrt->table, &p);
```

Comment:
this should be `route_node_lookup()` since we want to find the exact prefix that was given.